scripts/db/src/login.ts

```diff
@@ -10,54 +14,117 @@ async function login(
   try {
     /** Get required dependencies */
     const bcrypt = require('bcrypt@5.0.1') as { compare: typeof compare }
-    const { Client } = require('pg@7.17.1')
+    const { createHash } = require('crypto') as {
+      createHash: typeof createHash_
+    }
+    const { Client: PGClient } = require('pg@7.17.1')
+    const { MongoClient } = require('mongodb@3.1.4')
+
+    /**
+     * Logic in this function tries to match:
+     * EAForum/packages/lesswrong/server/vulcan-lib/apollo-server/authentication.tsx
+     */
+    async function loginForumUser(): Promise<CallbackUser | null> {
+      // Connect to the Forum DB
+      const mongoClient: MongoClient = new MongoClient(configuration.MONGO_URI)
+      await mongoClient.connect()
+
+      // Query the users collection for someone with our email
+      const forumUser = await mongoClient
+        .db(configuration.MONGO_DB_NAME)
+        .collection<ForumUser>('users')
+        .findOne({ 'emails.address': email })
+
+      await mongoClient.close()
+
+      if (!forumUser) {
+        return null
+      }
+
+      // Check their password
+      // Meteor hashed its passwords twice, once on the client and once again on
+      // the server. To preserve backwards compatibility with Meteor passwords,
+      // we do the same, but do it both on the server-side
+      const meteorClientSideHash = createHash('sha256')
+        .update(password)
+        .digest('hex')
+      const isValid = await bcrypt.compare(
+        meteorClientSideHash,
+        forumUser.services.password.bcrypt
+      )
+      if (!isValid) {
+        return null
+      }
 
-    /** Declare connection info */
-    const connectionInfo: ConnectionConfig = {
-      user: configuration.POSTGRES_USERNAME,
-      password: configuration.POSTGRES_PASSWORD,
-      host: configuration.POSTGRES_HOST,
-      database: configuration.POSTGRES_DATABASE,
-      port: configuration.POSTGRES_PORT
-        ? parseInt(configuration.POSTGRES_PORT)
-        : 5432,
-      ssl: TEMPLATE_DATA.pgShouldSsl,
+      return {
+        id: forumUser._id,
+        nickname: forumUser.displayName,
```

Comment:
How does this interact with other applications? I suspect it's just fine (provided we don't require usernames at the Auth0 tenant level), but just making sure this doesn't have any weird interactions.

Reply:
Yes, it is weird, and I have under-thought about it.

Reply:
Setting username. We'll want to make sure emails are set to try username when they send, for which I've added a TODO. See more info in slack #cea-tech.